### ISSUE TITLE
false positive unused require

### ISSUE THREAD

**Issue by johanatan** [borkdude/clj-kondo#235]
**version**

latest

**platform**

linux

**problem**

see title.

**repro**

```
(ns redacted.utils
  (:require [redacted.specs :as specs]))

(defn format-rule-row [{log-action (:id specs/log-action)   ;; clj-kondo doesn't detect specs/
                        response-action (:id specs/response-action)}] ;; ditto
  (when (= "ignore" log-action response-action)
    {:class "inactive"}))
```

**expected behavior**

I would expect clj-kondo to see that the aliased require `specs` is indeed being used.

**actual behavior**

It does not see that.

**Comment by borkdude** [comment]
@johanatan Is this even supposed to work in Clojure? When I try something like this, it seems it doesn't work?

```
(ns redacted.utils
  (:require [clojure.string :as str]))

(defn edge-case [{i str/includes?
                  s str/starts-with?}]
  [i s])

(println (edge-case {})) ;;=> prints [nil nil]
```

**Comment by jleonard-r7** [comment]
Hmm, that’s a good point. It seems this code is confused. It must be dead code so I’ll remove it instead. Sorry for the churn.

**Comment by johanatan** [comment]
Hmm, that’s a good point. It seems this code is confused. It must be dead code so I’ll remove it instead. Sorry for the churn.

**Comment by borkdude** [comment]
@johanatan Maybe clj-kondo should warn about the the right-hand side of the bindings can only be keywords. I should check if this is true though.

**Comment by borkdude** [comment]
@johanatan Looking deeper at this, this example seems to work:

```
(ns redacted.utils
  (:require [clojure.string :as str]))

(println
 (let [{i str/includes?
        s str/starts-with?} {str/includes? 1 str/starts-with? 2}]
   [i s]))
```

So in your example `(:id specs/log-action)` can return a keyword, which will in turn be used for associative destructuring. So it seems that this should be fixed after all.

**Comment by borkdude** [comment]
@johanatan You can test the fix with this binary:

https://3410-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.06.02-alpha-SNAPSHOT-linux-amd64.zip

**Comment by johanatan** [comment]
Any chance you could deploy that so `latest` would pick it up? We're using leiningen to invoke clj-kondo.

**Comment by johanatan** [comment]
And, btw, `(:id specs/log-action)` is indeed a keyword in our codebase (I just looked it up).

**Comment by borkdude** [comment]
@johanatan After testing, it will be merged to master and then it will be automatically deployed as a SNAPSHOT version.

**Comment by johanatan** [comment]
Ok, please let me know once that's done. Thx

**Comment by borkdude** [comment]
@johanatan Any chance you could test this manually? before I merge this fix to master? That's why I provided you the binary. I'd like to prevent unnecessary commits to master.

**Comment by borkdude** [comment]
Alternatively you can checkout [this branch](https://github.com/borkdude/clj-kondo/commits/false-positive-associative-destructuring) and `lein install` it, if you rather want to run it with leiningen.

**Comment by johanatan** [comment]
Looks like you added a test case. As long as a) that test case is something you want to support and b) this code actually fixed it and c) without regressing anything else, then you should be good to merge to master. 

If at some point down the line, I find that this does not resolve my issue, I’ll definitely get back to you.

**Comment by borkdude** [comment]
@johanatan It's available from clojars in this version now: https://clojars.org/clj-kondo/versions/2019.06.02-alpha-SNAPSHOT
I'll release a non-SNAPSHOT version probably this weekend. Thanks for reporting the issue.

**Comment by johanatan** [comment]
Yep, that version fixes the issue. Thx!

**Comment by johanatan** [comment]
I've found another permutation of this bug:

clj-kondo doesn't seem to be schema-aware. So the sole use of a required namespace within a Schema definition such as the following will not be detected:

```
(defn a-func [*errors       :- (s/atom (s/maybe validation/TranslatedError))] nil)
```

Would you rather me file a separate issue for that?

**Comment by johanatan** [comment]
And another one.

References to the items in the following `:refer` list are not recognized to be uses of their parent namespace:

```
[redacted.string :refer [substr? ?string trim trim-l-1r]]
```
